File: Bit_Manipulation/python/next_number.py
from numpy import uint32
from exceptions import *


def get_bit(num, i):
    return ((num & (1 << i)) != 0)


def set_bit(num, i):
    return num | (1 << i)


def clear_bit(num, i):
    mask = ~(1 << i)

    return num & mask


def get_ones_count(num):
    count = 0

    for i in range(32):
        if get_bit(num, i):
            count += 1

    return count


def get_next(num):
    trailing = True

    for i in range(32):
        if not get_bit(num, i):
            if not trailing:
                next = set_bit(num, i)
                break
        else:
            trailing = False

        if i == 31:
            raise NoOnesException(num)

    for i in range(32):
        if get_bit(num, i):
            next = clear_bit(next, i)
            return next


def get_prev1(num):
    for i in range(32):
        if get_bit(num, i):
            first_one_index = i
            prev = clear_bit(num, i)
            break

    for i in range(32):
        if i + 1 == first_one_index:
            prev = set_bit(prev, i)

    return prev


def get_prev2(num):
    ones = 0
    trailing = False

    for i in range(32):
        if get_bit(num, i) and trailing:
            index = i
            break
        elif get_bit(num, i):
            ones += 1
        else:
            trailing = True

    prev = num

    for i in range(32):
        if i == index:
            prev = clear_bit(prev, i)
            return prev
        elif index - ones - 1 <= i:
            prev = set_bit(prev, i)
        else:
            prev = clear_bit(prev, i)



def get_prev(num):
    if not get_bit(num, 0):
        return get_prev1(num)

    return get_prev2(num)


def main():
    num = uint32(369)
    print("{0:b}".format(num))

    # next = get_next(num)
    # print("{0:b}".format(next))

    prev = get_prev(num)
    print("{0:b}".format(prev))


if __name__ == '__main__':
    main()

File: Bit_Manipulation/go/next_number.go
package main


import (
  "fmt"
)


func getBit(num uint32, i int) bool {
  return ((num & (1 << i)) != 0)
}


func setBit(num uint32, i int) uint32 {
  return num | (1 << i)
}


func clearBit(num uint32, i int) uint32 {
  var mask uint32 = ^(1 << i)
  return num & mask
}


func getNext(num uint32) uint32 {
  var trailing bool = true
  var next uint32

  for i := 0; i < 32; i++ {
    if !getBit(num, i) {
      if !trailing {
        next = setBit(num, i)
        break
      }
    } else {
      trailing = false
    }

    if i == 31 {
      return 0
    }
  }

  for i := 0; i < 32; i++ {
    if getBit(num, i) {
      next = clearBit(next, i)
      return next
    }
  }

  return 0
}


func getPrev1(num uint32) uint32 {
  var prev uint32
  var firstOneIndex int

  for i := 0; i < 32; i++ {
    if getBit(num, i) {
      firstOneIndex = i
      prev = clearBit(num, i)
      break
    }
  }

  for i := 0; i < 32; i++ {
    if i + 1 == firstOneIndex{
      prev = setBit(prev, i)
    }
  }

  return prev
}


func getPrev2(num uint32) uint32 {
  var prev uint32 = num
  var ones, index int
  var trailing bool

  for i := 0; i < 32; i++ {
    if getBit(num, i) && trailing {
      index = i
      break
    } else if getBit(num, i) {
      ones++
    } else {
      trailing = true
    }
  }

  for i := 0; i < 32; i++ {
    if i == index {
      prev = clearBit(prev, i)
      break
    } else if index - ones - 1 <= i {
      prev = setBit(prev, i)
    } else {
      prev = clearBit(prev, i)
    }
  }

  return prev
}


func getPrev(num uint32) uint32 {
  if !getBit(num, 0) {
    return getPrev1(num)
  }

  return getPrev2(num)
}


func main() {
  var a, next, prev uint32
  a = 368

  fmt.Printf("%032b\n", a)

  next = getNext(a)
  fmt.Printf("%032b\n", next)

  prev = getPrev(a)
  fmt.Printf("%032b\n", prev)
}
